### DISCUSSION TITLE
datasets/obb/dota-v2/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#10142] [Docs]
# datasets/obb/dota-v2/

Delve into DOTA, an Oriented Bounding Box (OBB) aerial imagery dataset with 1.7 million instances and 11,268 images.

https://docs.ultralytics.com/datasets/obb/dota-v2/

**Comment by hencai** [comment]
how to test DOTA datatests online using YOLOv8-obb? somebody help me ?

**Reply by glenn-jocher** [reply]
Hey there! 😊 Testing the DOTA dataset online using YOLOv8-obb is quite straightforward. If you've already prepared your dataset according to the Ultralytics guidelines and split the images, here's a simple example to get you started with training:

```bash
# Train a YOLOv8n-OBB model on the DOTA v2 dataset
yolo obb train data=DOTAv1.yaml model=yolov8n-obb.pt epochs=100 imgsz=640
```

This command assumes you're using a version of the DOTA dataset (like DOTA v1 for simplicity) and you have a corresponding `.yaml` file (`DOTAv1.yaml`) describing your dataset paths and classes. Adjust `epochs` and `imgsz` based on your needs. 

Just make sure to check out the [DOTA dataset page](https://docs.ultralytics.com/datasets/obb/dota-v2/) on Ultralytics docs for more info on dataset preparation and additional configurations! Happy detecting! 🚀

**Comment by Tintin26100** [comment]
How can I find the number of instances per class in DOTA v1.5 ? I've been searching for this information for a moment and I can't find any answers

**Reply by pderrenger** [reply]
@Tintin26100 hello!

The number of instances per class in DOTA v1.5 isn't directly listed in the dataset's summary. However, you can obtain this information by analyzing the annotation files provided with the dataset. Each object instance in these files is labeled with a class, and you can count the number of times each class appears.

If you need further guidance on how to process these annotation files or any other assistance with the DOTA dataset, feel free to ask!

Happy coding! 😊

**Reply by Tintin26100** [reply]
Hi, thanks for your quick answer, I will gladly accept your help because I don't really know what to do to extract those informations, where do you find the annotation files ? Is it on the github ? If it is what's it's path ?
Thanks for the time you spend helping

**Reply by pderrenger** [reply]
@Tintin26100 hello!

The annotation files for the DOTA dataset are typically included within the dataset download itself. These files contain the details for each instance, including class labels and oriented bounding box coordinates.

For the DOTA v1.5 dataset, you can find these files in the dataset directory after you download it from the official DOTA dataset page or the provided download link in the dataset's YAML configuration file. The path to the annotations will usually be structured under the `labels` or `annotations` directory within the main dataset folder.

If you've already downloaded the dataset, you should look for `.txt` or `.json` files corresponding to each image in the dataset, which contain the annotations. If you need further assistance navigating these files or extracting specific data, feel free to ask!

Thank you for reaching out, and happy analyzing! 😊

**Comment by jack-ready** [comment]
Recently, I've been using YOLOv8 to train the DOTAv1 dataset and encountered some issues. I was wondering if I could seek advice. 
1.To train the DOTA dataset, we split the original DOTA images with high-resolution into images with 1024x1024. I noticed that the provided segmentation code provides three sizes, rates=[0.5, 1.0, 1.5]. In order to obtain 1024×1024 images, it seems that we should choose rates=1, but I noticed that some images in the segmentation results of the validation set are smaller than 1024×1024, such as 1024×1023, 824*1024, etc. What is the reason for this？Will it affect the experiment？
2.Another issue arises during the training process. I noticed that the training process conducts validation after each epoch. I attempted to disable validation by setting val=False, but it seems to have no effect. What could be the reason for this? The training code snippet is as follows: results = model.train(data="DOTAv1.yaml", epochs=100, imgsz=640, batch=10, workers=0, val=False). Additionally, during validation, it seems that the images are not being resized to 640. Printing the results in the code yields torch.Size([10, 3, 672, 672]), whereas during the training process, images are correctly resized to 640. Could someone help me understand this?

**Reply by glenn-jocher** [reply]
@jack-ready hi there!

Thank you for reaching out with your questions regarding training the DOTA dataset with YOLOv8. Let's address each of your concerns:

1. **Image Splitting and Sizes**:
   When splitting high-resolution DOTA images into 1024x1024 segments using `rates=[0.5, 1.0, 1.5]`, you might notice some resulting images are slightly smaller than 1024x1024 (e.g., 1024x1023 or 824x1024). This occurs due to the way the splitting algorithm handles the edges of the original images. If the original image dimensions are not perfectly divisible by 1024, the last segment in each row or column might be smaller. This should not significantly affect your experiment as long as the majority of your data maintains the expected size. However, you can ensure consistency by padding these smaller images to 1024x1024 if necessary.

2. **Disabling Validation**:
   The `val=False` argument should indeed disable validation during training. If it doesn't seem to work, please ensure you are using the latest version of the Ultralytics package. You can update it using:
   ```bash
   pip install --upgrade ultralytics
   ```
   If the issue persists, providing a minimum reproducible example would be very helpful for us to diagnose the problem. You can find guidance on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

3. **Image Resizing During Validation**:
   The discrepancy in image sizes during validation (e.g., `torch.Size([10, 3, 672, 672])` instead of 640) might be due to the model's internal resizing logic. Ensure that the `imgsz` parameter is correctly set for both training and validation. You can explicitly specify the validation image size in your training script:
   ```python
   results = model.train(data="DOTAv1.yaml", epochs=100, imgsz=640, batch=10, workers=0, val=True, val_imgsz=640)
   ```

I hope this helps! If you have any further questions or need additional assistance, feel free to ask. Happy training! 🚀

**Reply by jack-ready** [reply]
Thank you very much for your help. I have another question regarding the YOLOV8 training on the DOTA v1.0 dataset. I would like to know how to set the training parameters, such as learning rate, optimizer, and loss function. When I try to reproduce the results, my mAP50 is about 74%, but it does not reach 78%. I would greatly appreciate any assistance you can provide.Below is my training code.
from ultralytics import YOLO
model = YOLO("yolov8n-obb.yaml")
model.load("yolov8n-obb.pt")
model = YOLO("yolov8n-obb.yaml").load("yolov8n.pt") 
results = model.train(data="DOTAv1.yaml", epochs=100, imgsz=640, batch=48, workers=0,project="runs/DOTAV1.0",val=True,resume=True)

**Reply by glenn-jocher** [reply]
Hi there! Glad to help with your training setup. To fine-tune your training parameters for better mAP50 results, you can adjust the learning rate, optimizer, and loss function. For instance, you might try a lower learning rate or switch to a different optimizer like Adam. Here's a quick example:

```python
from ultralytics import YOLO

model = YOLO("yolov8n-obb.yaml").load("yolov8n-obb.pt")
results = model.train(
    data="DOTAv1.yaml", 
    epochs=100, 
    imgsz=640, 
    batch=48, 
    workers=0, 
    project="runs/DOTAV1.0", 
    val=True, 
    lr0=0.001,  # Adjust learning rate
    optimizer='Adam'  # Change optimizer
)
```

Ensure you're using the latest version of the Ultralytics package to avoid any bugs. If you still face issues, feel free to share more details. Happy training! 🚀

**Comment by XWP-PPP** [comment]
May I ask how you converted the inference results from YOLOV8 OBB into inference results that can be uploaded to DOTA's Task 1 server? I saved the final inference results of the model in the form of xyxyxy and uploaded them via email. However, the mAP results provided on YOLO's official website differ significantly. I would like to know the reason for this. Can you provide me with a conversion code to convert the test image set into a result that can be uploaded to DOTA's Task 1 server?
Thank you very much for your help!

**Reply by pderrenger** [reply]
@XWP-PPP hi there!

Thank you for your question. Converting YOLOv8 OBB inference results to a format suitable for DOTA's Task 1 server involves ensuring the results are in the correct format and structure expected by the DOTA evaluation server. Here’s a step-by-step guide to help you with this process:

1. **Ensure Correct Format**: DOTA's Task 1 server expects results in a specific format. Typically, this involves saving the results in a text file where each line corresponds to a detection in the format:
   ```
   <image_id> <class> <confidence> <x1> <y1> <x2> <y2> <x3> <y3> <x4> <y4>
   ```

2. **Conversion Code**: Below is a Python snippet to convert YOLOv8 OBB results to the required format:

    ```python
    from ultralytics import YOLO
    import numpy as np

    # Load your trained YOLOv8 OBB model
    model = YOLO("path/to/your/yolov8n-obb.pt")

    # Define the path to your test images
    test_images = "path/to/test/images"

    # Run inference
    results = model.predict(source=test_images, save=False)

    # Function to convert results to DOTA format
    def convert_to_dota_format(results, output_file):
        with open(output_file, 'w') as f:
            for result in results:
                image_id = result.path.stem
                for box in result.boxes:
                    class_id = int(box.cls)
                    confidence = box.conf
                    points = box.xyxyxyxy
                    points_str = ' '.join(map(str, points.flatten()))
                    f.write(f"{image_id} {class_id} {confidence} {points_str}\n")

    # Convert and save results
    convert_to_dota_format(results, "dota_results.txt")
    ```

3. **Upload to DOTA Server**: Once you have the results in the correct format, you can upload the `dota_results.txt` file to the DOTA Task 1 server via their specified submission method.

4. **Verify Results**: Ensure that the results are consistent with the expected format and that the image IDs match those in the test set. Discrepancies in mAP could arise from differences in how results are formatted or processed.

For more detailed information on the DOTA dataset and its structure, you can refer to the [DOTA Dataset Documentation](https://docs.ultralytics.com/datasets/obb/dota-v2/).

If you encounter any issues or have further questions, feel free to ask. Happy coding! 😊

**Comment by huynhkhoa44** [comment]
I have labeled data in YOLOv8 object detection format, our label data format is already rotated rectangles to the correct angle of the object, however since it is in YOLOv8 object detection format, the "angle" parameter is missing. I no longer keep the label folder on CVAT so I cannot export to a different label format. Is there a way to add the angle value from the 4 pre-existing <x_center> <y_center> <width> <height> values?

**Reply by glenn-jocher** [reply]
To add the angle value from your existing YOLOv8 labels, you'll need to calculate it based on the coordinates. Unfortunately, without the original CVAT labels, this requires manual computation or using a tool that supports angle annotations. For guidance on datasets with oriented bounding boxes, you can refer to the DOTA dataset documentation at https://docs.ultralytics.com/datasets/obb/dota-v2/.

**Comment by keeper-jie** [comment]
I have download dotav1 form `https://github.com/ultralytics/assets/releases/download/v0.0.0/DOTAv1.zip` and use script split dataset.
```
from ultralytics.data.split_dota import split_test, split_trainval

# split train and val set, with labels.
split_trainval(
    data_root="/dataset/coco/DOTAv1",
    save_dir="/dataset/coco/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
)
# split test set, without labels.
split_test(
    data_root="/dataset/coco/DOTAv1/",
    save_dir="/dataset/coco/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
)
```
how to reproduce yolov11 obb result, please show me the command like 
```
yolo train model=ultralytics-8.3.13/ultralytics/cfg/models/11/yolo11n-obb.yaml data=DOTAv1/DOTAv1.yaml epochs=500 batch=32*4 imgsz=640 device=4,5,6,7 cache=disk amp=True
```

**Reply by glenn-jocher** [reply]
@keeper-jie for training YOLO11-OBB on DOTAv1 after splitting, use this command with your 4 GPUs:  
`yolo obb train data=DOTAv1.yaml model=yolo11n-obb.pt epochs=500 imgsz=1024 batch=128 device=4,5,6,7`  

Key parameters from our DOTA documentation:  
- `imgsz=1024` (matches dataset's multiscale split)  
- Use official model name `yolo11n-obb.pt` rather than YAML path  
- Batch size 128 (32*4 GPUs) via `batch=128`  

Ensure your dataset YAML paths point to the split images directory.

**Comment by Taufiq-Ahmed-pranto** [comment]
Hi, I want to use the DOTA dataset only for YOLOv11 Nano. I don't want to use the model with oriented bounding boxes (OBB). What is the procedure?
